### ISSUE TITLE
Qwen 2.5 Base unable to generate <|im_end|> even after finetuning

### ISSUE THREAD

**Issue by MattYoon** [axolotl-ai-cloud/axolotl#2073]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Qwen 2.5 Base should generate <|im_end|> since I finetuned it on the official chat_template.

### Current behaviour

The finetuned base model does not generate <|im_end|> and generates gibberish after a certain point. I fine-tuned the model on MetaMathQA, and notice how even after generating `The answer is: 11 `, the model does not generate <|im_end|>.

`'<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nRoger has 5 tennis balls. He buys 2 more cans of tennis balls. Each can has 3 tennis balls. How many tennis balls does he have now?<|im_end|>\n<|im_start|>assistant\nRoger starts with 5 tennis balls.\nHe buys 2 more cans of tennis balls, and each can has 3 tennis balls, so he adds 2 * 3 = 6 tennis balls to his collection.\nIn total, Roger now has 5 + 6 = 11 tennis balls.\n#### 11\nThe answer is: 11 prostituerade\nThe answer is: 11%timeout\nThe correct answer is: 11 eskorte\nThe number of tennis balls Roger has now is the sum of the initial number of tennis balls (5) and the additional tennis balls he bought (6). So, 5 + 6 = 11 tennis balls.\nTherefore, Roger has 11 tennis balls in total.\n#### 11\nThe answer is: 11VRTX\nThe answer is: 11VRTX\nThe number of tennis balls Roger has now is the sum of the initial number of tennis balls (5) and the additional tennis balls he bought (6).\nSo, 5 + 6 = 11 tennis balls.\nTherefore, Roger has 11 tennis balls in total.\n#### 11\nThe answer is: 11VRTX\nThe number of tennis balls Roger has now is the'
`

### Steps to reproduce

I've attached the yaml file, but I believe any configuration will have the same issue.

### Config yaml

```yaml
base_model: Qwen/Qwen2.5-1.5B
trust_remote_code: false

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: pbatra/MetaMathQA-ShareGPT
    type: chat_template
    field_messages: conversations
    message_field_role: from
    message_field_content: value
    roles:
      user: ["human", "user"]
      assistant: ["gpt", "assistant", "ai"]
      system: ["system"]

dataset_prepared_path:
val_set_size: 0.0
output_dir: ./outputs/qwen-1.5b-metamath-100k

sequence_len: 2048
sample_packing: true
eval_sample_packing: true
pad_to_sequence_len: true

wandb_project: merge
wandb_entity: 
wandb_watch:
wandb_name: qwen-1.5b-metamath-100k
wandb_log_model:

gradient_accumulation_steps: 1
micro_batch_size: 128
num_epochs: 1
optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 3e-5

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: true

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_ratio: 0.05
saves_per_epoch: 2
evals_per_epoch: 0
debug:
weight_decay: 0.01
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<|im_start|>"
  eos_token: "<|im_end|>"
  pad_token: "<|endoftext|>"


plugins:
  - axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true
```

### Possible solution

If I use llama3's chat_template, there is no issue. The model is able to generate <|eot_id|><|endoftext|> successfully. 

I tried some attempts to make the official Qwen template work.
1. I thought maybe something was not right in the official configs. So I replaced all the configs of the base model with the Instruct model. --> Did not work.
2. I came across [this Tweet](https://twitter.com/danielhanchen/status/1856442699689414970) by Daniel Han. I tried his Unsloth version. --> Did not work.

Daniel Han says the official template should not be used for the base models since they are untrained, but I don't quite understand it. Even if it were untrained, shouldn't finetuning still increase the logprob of the tokens enough?

I'm not sure if this is a unfixable design flaw of Qwen 2.5 or a funny interaction with training frameworks that can be fixed on Axolotl's side.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main/71d4030b790d3d4ff9bae6bf3696cdc8e72bb5f7

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey, thanks for the report.

> The finetuned base model does not generate <|im_end|> and generates gibberish after a certain point. 

> Daniel Han says the official template should not be used for the base models since they are untrained, but I don't quite understand it. Even if it were untrained, shouldn't finetuning still increase the logprob of the tokens enough?

I noticed that you trained for one epoch only. Given that the base model has those tokens untrained, maybe a few more epochs may be a good step?

> If I use llama3's chat_template, there is no issue. The model is able to generate <|eot_id|><|endoftext|> successfully.

This is interesting. What happens if you use `chat_template: chatml`? The default chat_template uses `im_end` but in a slightly different manner.

Just a thought, it could also be that the initialization of new tokens are better than using ones in the model (which may be initialized poorly), although I haven't checked the model myself.

**Comment by ehartford** [comment]
Also - they say that <|endoftext|> doesn't work for the pad.

https://twitter.com/danielhanchen/status/1856442699689414970

**Comment by ehartford** [comment]
Try using this base model 
https://huggingface.co/unsloth/Qwen2.5-Coder-32B

**Comment by MattYoon** [comment]
@ehartford I tried the unsloth version. Specifically https://huggingface.co/unsloth/Qwen2.5-1.5B , did not work :(

**Comment by NanoCode012** [comment]
> Also - they say that <|endoftext|> doesn't work for the pad.
> 
> https://twitter.com/danielhanchen/status/1856442699689414970

I saw this as well, but did they say why?

**Comment by MattYoon** [comment]
@NanoCode012 
https://x.com/vancoykendall/status/1856524659824771106

His justification. But I agree with the person who asked the question in the link. If <|im_end|> was working in the first place, the pad token being the same as eos should not be an issue.

**Comment by winglian** [comment]
@MattYoon so here's the training I did https://wandb.ai/axolotl-ai/qwen-im-end/runs/uz5v11h9
using this config https://wandb.ai/axolotl-ai/qwen-im-end/runs/uz5v11h9/files/tmp/axolotl_config_g2g7wwya.yml

I'm not sure how you were testing inference, but I have PR  #2092 which fixes an issue where i was able to reproduce when using our inference code, but it was bugged and not using the chat template. 

when I tested using the build in inference with the fix and yaml above, I got

`CUDA_VISIBLE_DEVICES=0 python -m axolotl.cli.inference qwen.yaml --base_model="./outputs/qwen-1.5b-metamath-100k"`


```
================================================================================
Give me an instruction (Ctrl + D to submit): 
solve for x: x + 41 = x^2
========================================
<|im_start|>system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>
<|im_start|>user
solve for x: x + 41 = x^2<|im_end|>
<|im_start|>assistant
First, we subtract $x$ from both sides to get $41 = x^2 - x$.
Next, we rearrange the equation into standard quadratic form as $x^2 - x - 41 = 0$.
To solve this quadratic equation, we can factor it or use the quadratic formula.
Factoring, we find $(x-11)(x+3) = 0$, so $x=11$ or $x=-3$.
Using the quadratic formula, we have $x = \frac{-(-1)}{2} \pm \sqrt{\left(\frac{-(-1)}{2}\right)^2 - 4(1)(-41)} = \frac{1}{2} \pm \sqrt{1 + 684} = \frac{1}{2} \pm \sqrt{685}$.
Therefore, the solutions are $x=\boxed{\frac{1}{2}+\sqrt{685}, \frac{1}{2}-\sqrt{685}}$.The answer is: \frac{1}{2}+\sqrt{685},\frac{1}{2}-\sqrt{685}<|im_end|>
========================================
```
